### PR TITLE
3679 simplify cold chain graphing

### LINKED ISSUE #3679: Simplify Cold Chain Graphing
## The suggestion

Currently the cold chain graphs have been a significant source of bugs and confusion in open-mSupply.

I argue that a simpler graph view would serve the needs of the users better and reduce the confusion and complexity but would have some limitations.

I propose that chart should show every datapoint in the selected time range and use a dotted line to connect them.

Something similar to this...
https://master--63da8268a0da9970db6992aa.chromatic.com/?path=/docs/api-cartesian-line--docs

image

This is easy to understand, as you see all the datapoints from the system and the dotted line shows clearly that the gaps between the datapoints are interpolated not `real` data. The user can visually see if datapoints are further apart than previous ones indicating a gap, and can evaluate the severity of it.

To avoid the browser crashing with too much data, we'd need to limit the time range that the user can select.
This would mean if the user has a use case where they need to look at a long time range, they'd need to go to the dashboard to investigate (BTW: We know the dashboard can crash the browser by using too much memory when trying to draw too many datapoints on grafana too)

### Why should we invest time in this?

We've already sunk a lot of time in to the queries and graphs for cold chain data, yet bugs and confusion still abound.
See this for the latest: https://github.com/msupply-foundation/open-msupply/issues/3283

The current graphing approach isn't easy to understand for users, and developers need to keep checking data and logs to understand differences between the datapoints, the query logic, and the graphing view.

Alternative approaches with options to customise when something counts as a `gap` allowing the user to select different kinds of aggregations, etc all would result in a lot more work for us.

### Are there any risks associated with this change?

I think it reduces risks of mistakes and future bugs, but does potentially introduce new bugs with the updated code.

### How much effort is required?

This feels like around 1 day's work to me, but how we limit the time range might be a little bit complex?
We'll need to evaluate the risks of trying to display too much data, particularly on a tablet, but I think in general people understand if something is slow when trying to look at a lot of data...

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #3679 
Fixes [#3090](https://github.com/msupply-foundation/open-msupply/issues/3090)
Fixes [#3283](https://github.com/msupply-foundation/open-msupply/issues/3283)

# 👩🏻‍💻 What does this PR do?

>Note: To Compare the two graphs, check out branch `3679-simplify-graph-compare-deleteme`

The cold chain graph now uses the normal temperatureLog query to get data, and shows all datapoints instead of an averaged summary. If there are too many datapoints, not all data is displayed and a warning is shown.

This makes the graph display much more consistently, is easier to understand from a code perspective, and from the point of view of the users.

Old:
image

New:
image

Too much data...
image

## 💌 Any notes for the reviewer?

Code has ended up a little bit messy sorry! Can clean up but figured better get some review first. For example it might be better to split some of the helper functions into another file similar to the old `useTemperatureChartData.ts`
Or maybe we've already spent too much time and just want to merge?

It's probably worth creating a custom graphql request for the temperature logs that avoids using loaders for performance, was going to do that but ran out of time.

We also talked about making sensors an auto complete in the filter section to make it easier to find a particular sensor. Particularly if you have too much data to display, which could potentially mean we don't see the sensor in the list at all.

Would be nice to show when a breach ended as was as when it starts. Would be possible to do something like that in future.

> This removes graphql/API functionality which is generally not recommended. However, we have no other known consumers of the chart data, and we're trying to simplify here, so reducing code and complexity is a win IMHO

# 🧪 Testing

- [ ] Setup a datafile with cold chain data, and view the cold chain charts under Cold Chain -> Monitoring
- [ ] Experiment with different timeranges and make sure the data aligns with what's in the database.
- [ ] Check that all the existing filter still work.
- [ ] Test if you get an performance issues on Android..

# 📃 Documentation
- [ X ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour

### CODE REVIEW

**Comment by jmbrunskill** [comment]
FYI - Updated graph to include smaller dots based on review from @adamdewey

**Comment by roxy-dao** [comment]
Rebase to V2.0.0 please~

**Review by mark-prins — CHANGES_REQUESTED** [review]
🥇  That is brilliant! thanks James - the chart is super quick now, and doesn't redraw randomly. This feels a lot more solid than it was!

I have a [suggestion branch](https://github.com/msupply-foundation/open-msupply/compare/3679-simplify-cold-chain-graphing...3679-simplify-cold-chain-graphing(quick-tidy)) with the tweaks I made - the request for change is the imports, I don't remember anything critical in there.

I agree about lifting things out of the TemperatureChart file, at > 500 lines it's larger than I like - but agree that enough time has been spent. I've just lifted out the tooltip, the rest of it I think we should leave as it is now.

One small gripe is that you only get a tooltip for one sensor but I think we just live with that

Screenshot 2024-05-07 at 12 01 59 PM

Screenshot 2024-05-07 at 12 01 55 PM

doesn't matter which line is highest (temperature wise, that is) - only the first shows the tooltip if there are multiple lines.

File: client/packages/coldchain/src/Monitoring/ListView/TemperatureChart/TemperatureChart.tsx [hunk 1]
```diff
@@ -140,42 +299,87 @@ const Chart = ({
...
-      const to = customDate(new Date(), 'yyyy-MM-dd HH:mm');
-      updateQuery({ datetime: { from, to } });
-    }
-  }, []);
-
-  if (isLoading) {
-    return <BasicSpinner />;
-  }
-
-  if (!hasData) {
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/coldchain/src/Monitoring/ListView/TemperatureChart/TemperatureChart.tsx:158]: weird. not see that change now - your original had the no data message 🤷  don't know how I got that first image, but seems fine now.

**Comment by adamdewey** [comment]
> doesn't matter which line is highest (temperature wise, that is) - only the first shows the tooltip if there are multiple lines.

This seems to work ok for me - the tooltip only appears on the dots, so if you hover horizontally it will alternate between the lines depending on which dot it is next horizontally. 

At least, it works on tablets like that...

**Comment by mark-prins** [comment]
> This seems to work ok for me

I couldn't get it to show on the web version, didn't matter where I hovered it always showed the tip from the first sensor

**Comment by jmbrunskill** [comment]
> > This seems to work ok for me
> 
> I couldn't get it to show on the web version, didn't matter where I hovered it always showed the tip from the first sensor

I think it is trying to find the `closest point` to the cursor but does seem a bit weird. Looked into getting crosshairs to work but didn't find a good example to steal. Was thinking about finding the closest point for each line and display that (similar to the look up the old graph did, but I think we really only want to do that if the points from the lines are really close together. Could be done in future I think, maybe worth creating a follow up issue incase we want invest more time in future...

File: client/packages/coldchain/src/Monitoring/ListView/TemperatureChart/BreachIndicator.tsx [hunk 2]
```diff
@@ -11,26 +11,26 @@ export const BreachIndicator = ({
...
-          breach,
+          breachId,
           position: event.currentTarget.getBoundingClientRect(),
         })
       }
       x={cx - 13.5}
       y={cy - 13.5}
       fill={theme.palette.error.main}
-      sx={{ color: 'background.white', cursor: 'pointer' }}
+      sx={{ color: 'background.white', cursor: 'pointer', zIndex: 9999 }}
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply client/packages/coldchain/src/Monitoring/ListView/TemperatureChart/BreachIndicator.tsx]: Yeah found that and forgot to remove...

File: client/packages/coldchain/src/Monitoring/ListView/TemperatureChart/TemperatureChart.tsx [hunk 3]
```diff
@@ -17,104 +18,263 @@ import {
...
-import { BreachConfig, BreachDot, DotProps, Sensor } from './types';
+import { BreachDot, DotProps, ChartSeries } from './types';
 import { BreachIndicator } from './BreachIndicator';
 import { Toolbar } from '../TemperatureLog/Toolbar';
 import { useFormatTemperature } from '../../../common';
+import { TemperatureLogFragment } from '../../api';
+import { useTemperatureLogs } from '../../api/TemperatureLog/hooks/document/useTemperatureLogs';
+
+import { scaleTime } from 'd3-scale';
+import { Dot, XAxisProps } from 'recharts';
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply client/packages/coldchain/src/Monitoring/ListView/TemperatureChart/TemperatureChart.tsx]: I think this is done in your suggestion branch right? Anything more for me to do?

[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/coldchain/src/Monitoring/ListView/TemperatureChart/TemperatureChart.tsx]: no more to do thanks. I did the suggestion after the comment 😄

**Review by mark-prins — APPROVED** [review]
👍 Looks great! thanks James

**Comment by roxy-dao** [comment]
Please rebase before merging (:

**Comment by mark-prins** [comment]
this is me, trying to view the Q tag tooltip:


https://github.com/msupply-foundation/open-msupply/assets/9192912/f18fefca-f80a-4d37-9a4e-1a9539a8c060

**Pull request merged**